File: server/core/hermes_skills_bridge.py
"""
Hermes 技能 → desk-pet Behavior 系统桥接

Hermes 的 slash 技能系统（以 / 开头的命令）通过此模块
映射到 desk-pet 前端的 BehaviorRegistry 事件分发。

映射表定义在 data/hermes_skill_map.json 中，支持热加载。
"""
from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Any, Optional

log = logging.getLogger("hermes-skills-bridge")

# 默认技能映射表
_DEFAULT_SKILL_MAP: dict[str, str] = {
    "/help": "show_help",
    "/status": "show_status",
    "/emotion": "show_emotion",
    "/memory": "search_memory",
    "/clear": "clear_context",
    "/reset": "reset_personality",
    "/time": "show_time",
    "/dance": "trigger_dance",
    "/pet": "request_pet",
    "/sleep": "enter_sleep",
}


def _get_map_path() -> Path:
    """技能映射表路径：<desk-pet>/data/hermes_skill_map.json"""
    from core.session_service import get_session_db_path
    db_path = get_session_db_path()
    return db_path.parent / "hermes_skill_map.json"


def load_skill_map() -> dict[str, str]:
    """加载技能映射表。"""
    path = _get_map_path()
    if path.exists():
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            if isinstance(data, dict):
                return data
        except Exception as exc:
            log.warning("Failed to load skill map: %s", exc)
    return dict(_DEFAULT_SKILL_MAP)


def save_skill_map(mapping: dict[str, str]) -> None:
    """保存技能映射表。"""
    path = _get_map_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(mapping, ensure_ascii=False, indent=2), encoding="utf-8")
    log.info("Skill map saved: %d entries", len(mapping))


def detect_skill(text: str) -> Optional[tuple[str, str, str]]:
    """
    检测文本是否为技能调用。

    Returns:
        (完整命令, 技能名, 行为事件名) 或 None
    """
    text = text.strip()
    if not text.startswith("/"):
        return None

    # 提取命令（空格前的内容）
    parts = text.split(" ", 1)
    cmd = parts[0].lower()
    args = parts[1] if len(parts) > 1 else ""

    skill_map = load_skill_map()
    behavior_event = skill_map.get(cmd)

    if behavior_event:
        return (cmd, behavior_event, args)
    return None


def get_all_skills() -> list[dict]:
    """获取所有已注册的技能（含内置 + 自定义）。"""
    skill_map = load_skill_map()
    skills = []
    for cmd, behavior_event in skill_map.items():
        skills.append({
            "command": cmd,
            "behavior_event": behavior_event,
            "description": _get_skill_description(cmd, behavior_event),
            "is_builtin": cmd in _DEFAULT_SKILL_MAP,
        })
    return skills


def _get_skill_description(cmd: str, behavior_event: str) -> str:
    """获取技能描述。"""
    descriptions = {
        "/help": "显示帮助菜单",
        "/status": "查看宠物状态",
        "/emotion": "查看当前情绪",
        "/memory": "搜索记忆",
        "/clear": "清除对话上下文",
        "/reset": "重置人格状态",
        "/time": "查看当前时间",
        "/dance": "让宠物跳舞",
        "/pet": "请求抚摸",
        "/sleep": "进入睡眠模式",
    }
    return descriptions.get(cmd, f"触发 {behavior_event} 行为")

File: src/services/hermesSkillsBridge.ts
/**
 * Hermes 技能 → desk-pet Behavior 系统桥接
 *
 * 监听 Hermes Gateway 的 slash 技能事件（如 /help, /status），
 * 将其映射到 BehaviorRegistry 的事件分发系统。
 *
 * 技能映射表由服务端管理（data/hermes_skill_map.json），
 * 前端仅负责：检测技能调用 → 分发行为事件。
 */

import { createLogger } from '../utils/logger';
import { eventBus } from './eventBus';
import { getBehaviorRegistry } from './behavior/registry';
import { EventType } from './behavior/types';

const log = createLogger('HermesSkillsBridge');

export interface HermesSkill {
  command: string;
  behavior_event: string;
  description: string;
  is_builtin: boolean;
}

const SKILL_MAP_CACHE_KEY = 'deskpet_hermes_skill_map';

/**
 * 检测文本是否为技能调用，如果是则通过 BehaviorRegistry 分发。
 * 在 Hermes Gateway 的 onToken 回调中调用。
 *
 * @returns 是否匹配到技能
 */
export function detectAndDispatchSkill(text: string): boolean {
  const trimmed = text.trim().toLowerCase();

  // 只处理以 / 开头的消息
  if (!trimmed.startsWith('/')) return false;

  const cmd = trimmed.split(' ')[0];
  const args = trimmed.slice(cmd.length).trim();

  // 从缓存加载技能映射
  const skillMap = loadSkillMapCache();
  const behaviorEvent = skillMap[cmd];

  if (!behaviorEvent) return false;

  log.info('Skill detected & dispatched', { command: cmd, behaviorEvent, args });

  // 通过事件总线 + BehaviorRegistry 分发
  const registry = getBehaviorRegistry();

  // 根据行为事件类型映射到 EventType
  const eventType = mapBehaviorToEventType(behaviorEvent);

  if (eventType) {
    registry.dispatch(eventType, { command: cmd, args, behaviorEvent }).catch((err) => {
      log.warn('Skill dispatch failed', { err: String(err) });
    });
  }

  // 同时通过事件总线发射，方便其他模块监听
  eventBus.emit('hermes:skill', { command: cmd, args, behaviorEvent });

  return true;
}

/**
 * 从服务端加载技能列表（通过 Hermes Gateway REST API）
 */
export async function fetchSkillsFromServer(): Promise<HermesSkill[]> {
  try {
    const res = await fetch('http://127.0.0.1:8765/api/gateway/skills', {
      signal: AbortSignal.timeout(3000),
    });
    if (!res.ok) return [];
    const skills = (await res.json()) as HermesSkill[];
    // 更新缓存
    cacheSkillMap(skills);
    return skills;
  } catch {
    log.warn('Failed to fetch skills from server');
    return [];
  }
}

// ===== 内部 =====

function loadSkillMapCache(): Record<string, string> {
  try {
    const raw = localStorage.getItem(SKILL_MAP_CACHE_KEY);
    return raw ? (JSON.parse(raw) as Record<string, string>) : {};
  } catch {
    return {};
  }
}

function cacheSkillMap(skills: HermesSkill[]): void {
  const map: Record<string, string> = {};
  for (const s of skills) {
    map[s.command] = s.behavior_event;
  }
  try {
    localStorage.setItem(SKILL_MAP_CACHE_KEY, JSON.stringify(map));
  } catch {
    /* ignore */
  }
}

function mapBehaviorToEventType(behaviorEvent: string): EventType | null {
  const mapping: Record<string, EventType> = {
    show_help: EventType.INIT,
    show_status: EventType.INIT,
    show_emotion: EventType.EMOTION_CHANGED,
    search_memory: EventType.MESSAGE_SENT,
    clear_context: EventType.MESSAGE_SENT,
    reset_personality: EventType.PERSONA_CHANGED,
    show_time: EventType.TIMER,
    trigger_dance: EventType.DOUBLE_CLICK,
    request_pet: EventType.INTERACTION_PAT,
    enter_sleep: EventType.IDLE,
  };
  return mapping[behaviorEvent] ?? null;
}
